### PR TITLE
Feature: conditional check. Close #2927

### LINKED ISSUE #2927: Core : Conditionnal check
For my use case testing a Vaadin app, I need a mean to perform a check only if one condition against the response or the session is True.

I ended up by developping a check wrapper which take a function : (session, response) => boolean and a nested check in parameter and perform the nested check if the function return true.
I am not sure it has no impact on performance, but I need it cause my scenarios are highly coupled with the server responses.

Here is a simple use case for this feature : I check all responses with a jsonpath check, but some responses payload are not formated in json. So I supplied a jsonSecured check wich is a ConditionnalCheck with a condition on the reponse content-type header, to perform the jsonpath nested check only if the response contains a JSON payload.

Are you interested in this feature ?

### PR DESCRIPTION
**Pull request by mxbossard**
This PR replace previous #2931 . "checkIfOrElse" is fallen for sake of check scopes.

Add ConditionalCheck feature with checkIf only.
Add it's HttpCheck wrapper.
Add Unit Test.

I just dev it, I didn't use it for now. I hope the adaptations from previous PR are OK.
The Unit Test should keep us safe.

### CODE REVIEW

**Comment by mxbossard** [comment]
This PR replace previous #2931 . "checkIfOrElse" is fallen for sake of check scopes.

**Comment by mxbossard** [comment]
Hi @slandelle, 
I ping you cause I'm not sure you saw this is a new PR.

**Comment by slandelle** [comment]
@mxbossard I did saw it, but you know how things are when you return from vacation :) Will review later this week. Thanks!

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: Remove tons of empty lines

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+object ConditionalCheck {
+  
+  implicit def conditionalCheckBuilder2Check[R, C <: Check[R]](conditionalCheckBuilder: ConditionalCheckBuilder[R, C]): ConditionalCheck[R, C] = conditionalCheckBuilder.build
+
+  def checkIf[R, C <: Check[R]](condition: Expression[Boolean])(thenCheck: C)(implicit cw: ConditionalCheckWrapper[R, C]): C = cw.wrap(ConditionalCheckBuilder((r: R, s: Session) => condition(s), thenCheck))
+
+  def checkIf[R, C <: Check[R]](condition: (R, Session) => Validation[Boolean])(thenCheck: C)(implicit cw: ConditionalCheckWrapper[R, C]): C = 
+    cw.wrap(ConditionalCheckBuilder((r: R,s: Session) => condition(r, s), thenCheck))
+
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: kill empty line

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+package io.gatling.core.check
+
+import java.util.UUID
+import com.typesafe.scalalogging.StrictLogging
+import io.gatling.commons.validation._
+import io.gatling.core.session.{ Expression, Session }
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: Use Gatling import rules: java, empty line, scala, empty line, io.gatling, empty lines, others

File: gatling-http/src/test/scala/io/gatling/http/HttpCompileTest.scala [hunk 4]
```diff
@@ -1,284 +1,299 @@
...
+
+  val JSON_CONTENT_TYPE_VALUE: String = "application/json"
+
+  def isJsonResponse(response: Response): Boolean = {
+    response.isReceived && response.header(HttpHeaderNames.ContentType).exists { x => x.contains(JSON_CONTENT_TYPE_VALUE) }
+  }
+
+  def securedJsonCheck(check: HttpCheck): HttpCheck = ConditionalCheck.checkIf((response: Response, session: Session) => Success(isJsonResponse(response)))(check)
+
+}
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/HttpCompileTest.scala]: Why is all the file modified?! Different eol? Please fix.

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala (hunk 2 above)
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: This should go in CheckSupport so it's automatically available without requiring additional import.

File: gatling-http/src/test/scala/io/gatling/http/HttpCompileTest.scala [hunk 5]
```diff
@@ -1,284 +1,299 @@
...
+    .exponentialPauses
+    .uniformPauses(1.5)
+    .uniformPauses(1337 seconds)
+
+  import io.gatling.http.check.HttpCheck
+  import io.gatling.core.check.ConditionalCheck
+  import io.gatling.core.session.Session
+  import io.gatling.commons.validation.Success
+
+  val JSON_CONTENT_TYPE_VALUE: String = "application/json"
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/HttpCompileTest.scala]: Already available as `HttpHeaderValues.ApplicationJson`.
Plus, you use the Java convention for constants, not the Scala one :)

File: gatling-http/src/test/scala/io/gatling/http/HttpCompileTest.scala [hunk 6]
```diff
@@ -1,284 +1,299 @@
...
+
+  def isJsonResponse(response: Response): Boolean = {
+    response.isReceived && response.header(HttpHeaderNames.ContentType).exists { x => x.contains(JSON_CONTENT_TYPE_VALUE) }
+  }
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/HttpCompileTest.scala]: No need for braces as there's only one instruction

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  def wrap(check: ConditionalCheck[R, C]): C
+}
+
+case class ConditionalCheckBuilder[R, C <: Check[R]](condition: (R, Session) => Validation[Boolean], thenCheck: C) {
+  def build = ConditionalCheck[R, C](condition, thenCheck)
+}
+
+case class ConditionalCheck[R, C <: Check[R]](condition: (R, Session) => Validation[Boolean], thenCheck: C) extends Check[R] with StrictLogging {
+
+  val checkUuid: String = UUID.randomUUID.toString
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: private and lazy

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  
+  def performNestedCheck(nestedCheck: Check[R], response: R, session: Session)(implicit cache: scala.collection.mutable.Map[Any, Any]): Validation[CheckResult] = {
+    nestedCheck.check(response, session)
+  }
+  
+  def check(response: R, session: Session)(implicit cache: scala.collection.mutable.Map[Any, Any]): Validation[CheckResult] = {
+    val validationResult = condition(response, session).flatMap { c => 
+      if (c) {
+        logger.trace("Check: [{}] condition evaluate to true: perform nested then check.", checkUuid)
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: Please remove

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        performNestedCheck(thenCheck, response, session)
+      } else {
+        logger.trace("Check: [{}] condition is false: do not perform any check.", checkUuid)
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: Please remove

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala (hunk 7 above)
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: Please remove

File: gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+case class ConditionalCheck[R, C <: Check[R]](condition: (R, Session) => Validation[Boolean], thenCheck: C) extends Check[R] with StrictLogging {
+
+  val checkUuid: String = UUID.randomUUID.toString
+  
+  def performNestedCheck(nestedCheck: Check[R], response: R, session: Session)(implicit cache: scala.collection.mutable.Map[Any, Any]): Validation[CheckResult] = {
+    nestedCheck.check(response, session)
+  }
+  
+  def check(response: R, session: Session)(implicit cache: scala.collection.mutable.Map[Any, Any]): Validation[CheckResult] = {
+    val validationResult = condition(response, session).flatMap { c =>
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/check/ConditionalCheck.scala]: Please remove intermediate assignment and directly return (plus remove useless curly braces now there's one single chain.

File: gatling-http/src/main/scala/io/gatling/http/Predef.scala [hunk 11]
```diff
@@ -1,22 +1,34 @@
...
+
+object Predef extends HttpDsl {
+
+  type Request = org.asynchttpclient.Request
+  type Response = io.gatling.http.response.Response
+
+  implicit object HttpConditionalCheckWrapper extends ConditionalCheckWrapper[Response, HttpCheck] {
+    def wrap(check: ConditionalCheck[Response, HttpCheck]) = {
+      HttpCheck(check, check.thenCheck.scope, check.thenCheck.responseBodyUsageStrategy)
+    }
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/Predef.scala]: Remove useless curly braces

File: gatling-http/src/main/scala/io/gatling/http/Predef.scala [hunk 12]
```diff
@@ -1,22 +1,34 @@
...
+import io.gatling.core.check.ConditionalCheckWrapper
+import io.gatling.core.check.ConditionalCheck
+
+object Predef extends HttpDsl {
+
+  type Request = org.asynchttpclient.Request
+  type Response = io.gatling.http.response.Response
+
+  implicit object HttpConditionalCheckWrapper extends ConditionalCheckWrapper[Response, HttpCheck] {
+    def wrap(check: ConditionalCheck[Response, HttpCheck]) = {
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/Predef.scala]: missing `override` keyword

File: gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+import io.gatling.core.session.Session
+import io.gatling.http.HttpDsl
+import io.gatling.http.response.{ StringResponseBody, Response }
+
+import scala.collection.mutable
+
+import io.gatling.commons.validation.Success
+import io.gatling.core.check.ConditionalCheck._
+import io.gatling.http.Predef._
+import io.gatling.http.check.HttpCheck
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala]: Fix imports order

**Comment by slandelle** [comment]
@mxbossard Looks promissing. Could you please address comments and rebase into a single commit?

**Comment by slandelle** [comment]
@mxbossard gentle ping :)

**Comment by mxbossard** [comment]
Hi @slandelle, isn't it all ok ?
I think I adress all the comments.
We may just need some documentation somewhere.

**Comment by slandelle** [comment]
OK, the page was in my cache and didn't properly refresh and didn't see your changes! :(

One last thing: would it be possible for you to document this new feature, please? see https://github.com/gatling/gatling/blob/master/src/sphinx/http/http_check.rst

**Comment by mxbossard** [comment]
Last time I pinged you, I should have do it again :D 
I'll document this soon.

File: gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+import scala.collection.mutable
+
+import io.gatling.{ ValidationValues, BaseSpec }
+import io.gatling.core.CoreDsl
+import io.gatling.core.check.CheckResult
+import io.gatling.core.config.GatlingConfiguration
+import io.gatling.core.session.Session
+import io.gatling.http.HttpDsl
+import io.gatling.http.response.{ StringResponseBody, Response }
+import io.gatling.commons.validation.Success
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala]: misplaced

File: gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+import io.gatling.core.check.ConditionalCheck._
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala]: misplaced

**Comment by slandelle** [comment]
2 more nits :)

**Comment by slandelle** [comment]
@mxbossard Just to let you know: the plan is to release 2.2.1 tomorrow. I'm not pressuring you to catch this train, it can perfectly catch the next one, no worries.

File: gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala (hunk 15 above)
Comment (mxbossard) [inline gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala]: I don't understand what is misplaced ?
You want to me to sort import in alphabetical order ? Is that it ?

Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/check/body/ConditionalCheckSpec.scala]: Exactly :)

**Comment by mxbossard** [comment]
@slandelle Done.

**Comment by slandelle** [comment]
Cherry-picked (plus I formatted test code, you probably didn't compile it with test:compile).

Awesome work, thanks!

**Comment by slandelle** [comment]
cherry-picked in 56cc66a2cd650edb392fb9da8963224db44922ac

**Pull request closed without merging**